### ISSUE TITLE
TopologyException from version 2.2.0

### ISSUE THREAD

**Issue by eaalto** [MikkoVihtakari/ggOceanMaps#40]
After upgrading to the latest version (2.2.0), I am getting the following error:

Error in (function (msg) :
TopologyException: Input geom 0 is invalid: Self-intersection at -178.09999999999999 89.450000000000003

following a call to:
 basemap(limits=c(-105, 5, -5, 75), bathymetry=TRUE, bathy.style="contour_grey")

Traceback: 

3. basemap_data(limits = limits, data = data, shapefiles = shapefiles,
crs = crs, bathymetry = bathymetry, bathy.type = bathy.type,
downsample = downsample, glaciers = glaciers, lon.interval = lon.interval,
lat.interval = lat.interval, rotate = rotate, expand.factor = expand.factor, ...

4. basemap_data_crop(x = x, bathymetry = bathymetry, glaciers = glaciers,
crs = crs)

5. clip_shapefile(sf::st_transform(x$shapefiles$bathy, crs = x$crs),
limits = x$clip_limits)v

There are calls to 'sf' below that, but I checked and updating to the newest version of 'sf' (v 1.0.15) is ok. The problem goes away if I roll back to ggOceanMaps v. 2.1.1.

I'm not the familiar with how 'sf' and other spatial data works, so I don't have any suggestions about what would cause a TopologyException.

Thanks.
   Emil Aalto

**Comment by MikkoVihtakari** [comment]
Yes, seems like a bug somewhere. I'll look at it when I have time. Thanks for reporting.

**Comment by jmlondon** [comment]
Also getting a Topology Exception error in ggOceanMaps version 2.2.0. It seems to occur any time the default land layer is used AND a projection/plot that crosses the antimeridian is use. I can avoid the issue if I specify, for instance, `shapefiles = "Arctic"`

example data

```r
library(ggOceanMaps)
library(sf)
library(tidyverse)
library(rnaturalearth)

ak_base <- ne_states(
  country = 'United States of America',
  returnclass = "sf"
) |>
  dplyr::filter(name == "Alaska") |>
  st_transform(3571)

wa_base <- ne_states(
  country = 'United States of America',
  returnclass = "sf"
) |>
  dplyr::filter(name == "Washington") |>
  st_transform(3571)
```

Now, let's create an example without specifying a projection

```r
basemap(
  data = st_union(ak_base, wa_base),
  expand.factor = 1.25
) +
  geom_sf(data = wa_base) +
  geom_sf(data = ak_base)
```

![Image](https://github.com/user-attachments/assets/19013161-9fe2-40b5-adde-e3a1311f4303)

That seems to rely on the polar stereo projection but we would want to set the crs to 3571 

```r
basemap(
  data = st_union(ak_base, wa_base),
  expand.factor = 1.25, crs = 3571
) +
  geom_sf(data = wa_base) +
  geom_sf(data = ak_base)
```

This results in the following error:

```
! TopologyException: side location conflict at 164.1371295896916 42.867783060101907. This can occur if the input geometry is invalid.
```

If we specify use of the "Arctic" shapefile, then there's no TopologyException

```r
basemap(
  data = st_union(ak_base, wa_base),
  expand.factor = 1.25, crs = 3571,
  shapefiles = "Arctic"
) +
  geom_sf(data = wa_base) +
  geom_sf(data = ak_base)
```

![Image](https://github.com/user-attachments/assets/a3fc202a-d3b1-4638-abbb-a21478eb327d)

So, I'm guessing there must be some underlying issue with the default land geometry that's not an issue for the Arctic layer.

To demonstrate the issue is possibly linked to the antimeridian, if I plot only the Washington state region, there's no Topology Exception

```r
basemap(
  data = wa_base,
  expand.factor = 1.25, crs = 3571
) +
  geom_sf(data = wa_base)
```

![Image](https://github.com/user-attachments/assets/b56d9639-0aee-413f-9972-99f21862c94e)

But, when plotting the Alaska region alone (which crosses the antimeridian, the Topology Exception returns

```r
basemap(
  data = ak_base,
  expand.factor = 1.25, crs = 3571
) +
  geom_sf(data = ak_base)
```

```
! TopologyException: side location conflict at -132.48332960230908 44.137508267211857. This can occur if the input geometry is invalid.
```

**Comment by MikkoVihtakari** [comment]
It is true that the antimeridian conversion issue is complex and has not been entirely satisfactorily resolved. 

@jmlondon, I think your issue is not related to the original bug report. The original issue stems from a topology problem in the bathymetry contour shape file (https://github.com/MikkoVihtakari/ggOceanMapsLargeData/raw/master/data/dd_bathy.rda), as it does not cross the antimeridian. I'll need to play around with that shapefile to try to recreate it. I have not found time to do that yet. The workaround is to use raster bathymetry for now:

```r
basemap(limits=c(-105, 5, -5, 75), bathymetry=TRUE, bathy.style="rbg")
```

Your issue stems from trying to plot data using the decimal degree system for which the antimeridian is not defined: https://epsg.io/4326

Shapefile "Arctic" uses Arctic Polar Stereographic projection ([CRS 3995](https://mikkovihtakari.github.io/ggOceanMaps/reference/basemap.html)) with a defined antimeridian because the grid is meters from the North Pole. `basemap()` tries to avoid the undefined antimeridian issue in 4326 by using 3995 if coordinates are >30N. Hence, it works in your example, but if you st_transform() to 4326, you'll see the issue. 

I think the topology issue in your latest example stems from st_transform() making invalid geometries (similar to the original bug report) during the conversion from 4326 to 3571 (basemap(ak_base, expand.factor = 1.25) works; so does basemap(ak_base, expand.factor = 1.25, shapefiles = "Arctic", crs = 3571)). It may stem from the antimeridian, but these issues seem somewhat random. If you adjust the limits somewhat, it suddenly works again.

**Comment by MikkoVihtakari** [comment]
This should hopefully fix the issue: https://github.com/MikkoVihtakari/ggOceanMaps/commit/107c9ad5e21515d9857ed888709256decc4c375c

It makes maps with topology exceptions slightly slower to plot than maps that work without problems, but it worked at least with the example case.
